File: easypermissions/src/main/java/pub/devrel/easypermissions/RationaleDialogConfig.java
package pub.devrel.easypermissions;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.StyleRes;
import androidx.appcompat.app.AlertDialog;

/**
 * Configuration for either {@link RationaleDialogFragment} or {@link RationaleDialogFragmentCompat}.
 */
class RationaleDialogConfig {

    private static final String KEY_POSITIVE_BUTTON = "positiveButton";
    private static final String KEY_NEGATIVE_BUTTON = "negativeButton";
    private static final String KEY_RATIONALE_MESSAGE = "rationaleMsg";
    private static final String KEY_THEME = "theme";
    private static final String KEY_REQUEST_CODE = "requestCode";
    private static final String KEY_PERMISSIONS = "permissions";

    String positiveButton;
    String negativeButton;
    int theme;
    int requestCode;
    String rationaleMsg;
    String[] permissions;

    RationaleDialogConfig(@NonNull String positiveButton,
                          @NonNull String negativeButton,
                          @NonNull String rationaleMsg,
                          @StyleRes int theme,
                          int requestCode,
                          @NonNull String[] permissions) {

        this.positiveButton = positiveButton;
        this.negativeButton = negativeButton;
        this.rationaleMsg = rationaleMsg;
        this.theme = theme;
        this.requestCode = requestCode;
        this.permissions = permissions;
    }

    RationaleDialogConfig(Bundle bundle) {
        positiveButton = bundle.getString(KEY_POSITIVE_BUTTON);
        negativeButton = bundle.getString(KEY_NEGATIVE_BUTTON);
        rationaleMsg = bundle.getString(KEY_RATIONALE_MESSAGE);
        theme = bundle.getInt(KEY_THEME);
        requestCode = bundle.getInt(KEY_REQUEST_CODE);
        permissions = bundle.getStringArray(KEY_PERMISSIONS);
    }

    Bundle toBundle() {
        Bundle bundle = new Bundle();
        bundle.putString(KEY_POSITIVE_BUTTON, positiveButton);
        bundle.putString(KEY_NEGATIVE_BUTTON, negativeButton);
        bundle.putString(KEY_RATIONALE_MESSAGE, rationaleMsg);
        bundle.putInt(KEY_THEME, theme);
        bundle.putInt(KEY_REQUEST_CODE, requestCode);
        bundle.putStringArray(KEY_PERMISSIONS, permissions);

        return bundle;
    }

    AlertDialog createSupportDialog(Context context, Dialog.OnClickListener listener) {
        AlertDialog.Builder builder;
        if (theme > 0) {
            builder = new AlertDialog.Builder(context, theme);
        } else {
            builder = new AlertDialog.Builder(context);
        }
        return builder
                .setCancelable(false)
                .setPositiveButton(positiveButton, listener)
                .setNegativeButton(negativeButton, listener)
                .setMessage(rationaleMsg)
                .create();
    }

    android.app.AlertDialog createFrameworkDialog(Context context, Dialog.OnClickListener listener) {
        android.app.AlertDialog.Builder builder;
        if (theme > 0) {
            builder = new android.app.AlertDialog.Builder(context, theme);
        } else {
            builder = new android.app.AlertDialog.Builder(context);
        }
        return builder
                .setCancelable(false)
                .setPositiveButton(positiveButton, listener)
                .setNegativeButton(negativeButton, listener)
                .setMessage(rationaleMsg)
                .create();
    }

}

File: easypermissions/src/main/java/pub/devrel/easypermissions/dialogs/rationale/RationaleDialogConfig.kt
package pub.devrel.easypermissions.dialogs.rationale

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import androidx.annotation.StyleRes
import androidx.appcompat.app.AlertDialog
import pub.devrel.easypermissions.models.PermissionRequest
import java.util.ArrayList

private const val KEY_POSITIVE_BUTTON = "positiveButton"
private const val KEY_NEGATIVE_BUTTON = "negativeButton"
private const val KEY_RATIONALE_MESSAGE = "rationaleMsg"
private const val KEY_THEME = "theme"
private const val KEY_REQUEST_CODE = "requestCode"
private const val KEY_PERMISSIONS = "permissions"

/**
 * Configuration for either [RationaleDialogFragment] or [RationaleDialogFragmentCompat].
 */
internal class RationaleDialogConfig {

    var positiveButton: String? = null
    var negativeButton: String? = null
    var theme: Int = 0
    var requestCode: Int = 0
    var rationaleMsg: String? = null
    var permissions: List<String>? = null

    constructor(permissionRequest: PermissionRequest) {
        this.positiveButton = permissionRequest.positiveButtonText
        this.negativeButton = permissionRequest.negativeButtonText
        this.rationaleMsg = permissionRequest.rationale
        this.theme = permissionRequest.theme
        this.requestCode = permissionRequest.code
        this.permissions = permissionRequest.perms
    }

    constructor(bundle: Bundle) {
        positiveButton = bundle.getString(KEY_POSITIVE_BUTTON)
        negativeButton = bundle.getString(KEY_NEGATIVE_BUTTON)
        rationaleMsg = bundle.getString(KEY_RATIONALE_MESSAGE)
        theme = bundle.getInt(KEY_THEME)
        requestCode = bundle.getInt(KEY_REQUEST_CODE)
        permissions = bundle.getStringArrayList(KEY_PERMISSIONS)
    }

    fun toBundle(): Bundle {
        val bundle = Bundle()
        bundle.putString(KEY_POSITIVE_BUTTON, positiveButton)
        bundle.putString(KEY_NEGATIVE_BUTTON, negativeButton)
        bundle.putString(KEY_RATIONALE_MESSAGE, rationaleMsg)
        bundle.putInt(KEY_THEME, theme)
        bundle.putInt(KEY_REQUEST_CODE, requestCode)
        bundle.putStringArrayList(KEY_PERMISSIONS, permissions as ArrayList<String>?)

        return bundle
    }

    fun createSupportDialog(context: Context, listener: DialogInterface.OnClickListener): AlertDialog {
        val builder: AlertDialog.Builder
        if (theme > 0) {
            builder = AlertDialog.Builder(context, theme)
        } else {
            builder = AlertDialog.Builder(context)
        }
        return builder
            .setCancelable(false)
            .setPositiveButton(positiveButton, listener)
            .setNegativeButton(negativeButton, listener)
            .setMessage(rationaleMsg)
            .create()
    }

    fun createFrameworkDialog(
        context: Context,
        listener: DialogInterface.OnClickListener
    ): android.app.AlertDialog {
        val builder: android.app.AlertDialog.Builder
        if (theme > 0) {
            builder = android.app.AlertDialog.Builder(context, theme)
        } else {
            builder = android.app.AlertDialog.Builder(context)
        }
        return builder
            .setCancelable(false)
            .setPositiveButton(positiveButton, listener)
            .setNegativeButton(negativeButton, listener)
            .setMessage(rationaleMsg)
            .create()
    }
}